src/backend/distributed/shared_library_init.c

```diff
@@ -169,6 +172,17 @@ _PG_init(void)
 								"shared_preload_libraries.")));
 	}
 
+	/*
+	 * Stack size increase during high memory load may cause unexpected crashes.
+	 * With this alloca call, we are increasing stack size explicitly, so that if
+	 * it is not possible to increase stack size, we will get an OOM error instead
+	 * of a crash. Allocated memory will be automatically released at the end of
```

Comment:
To make is a bit more clear maybe: `Allocated memory will be automatically released at the end this function's scope.` -> 

```
This function is called on backend startup. The allocated memory will 
automatically be released at the end of the function's scope. However, 
we'd have already expanded the stack and it wouldn't shrink back. So, 
in a sense, per backend we're securing max_stack_depth kB's of 
memory on the stack upfront.

Not all the backends require max_stack_depth kB's on the stack, so we 
may unnecessary allocate. However, the default value is 2MB, which 
seems an acceptable trade-off. Also, allocating memory upfront may 
perform better under some circumstances. 
```

Reply:
The OpenBSD [alloca(3)](https://man.openbsd.org/alloca) man page warns:

> The alloca() function is unsafe because it cannot ensure that the pointer returned points to a valid and usable block of memory. The allocation made may exceed the bounds of the stack, or even go further into other objects in memory, and alloca() cannot determine such an error. Avoid alloca() with large unbounded allocations.

Perhaps C99's variable length array feature works more cleanly? I just don't know how you check for allocation failure. On my mac the following code prints up to 9223372036854775808 and then overflows back to 0 without any indication of failure along the way.

```c
/* compiled with -std=c99 */

#include <stdio.h>

void vla(size_t n)
{
	char memory[n];
	printf("%llu\n",
		(unsigned long long) sizeof(memory)/sizeof(memory[0]));
}

int main(void)
{
	size_t bytes = 1;
	while (bytes)
		vla(bytes = bytes * 2);
}
```